source-sftp/main.go

```diff
@@ -0,0 +1,397 @@
+	})
+
+	l.stack = append(l.stack, visits...)
+
+	return nil
+}
+
+// visit wraps fs.FileInfo and retains the full path to the file, since fs.FileInfo only provides
+// the file name.
+type visit struct {
+	fs.FileInfo
+	path string
+}
+
+func (s *sftpSource) Read(_ context.Context, obj filesource.ObjectInfo) (io.ReadCloser, filesource.ObjectInfo, error) {
+	file, err := s.client.Open(obj.Path)
+	if err != nil {
+		return nil, obj, fmt.Errorf("opening path %s for reading: %w", obj.Path, err)
+	}
+
+	// Update file info in case it has changed since the original listing.
+	info, err := file.Stat()
+	if err != nil {
+		return nil, obj, fmt.Errorf("stat'ing path %s for reading: %w", obj.Path, err)
+	}
+	obj.Size = info.Size()
+	obj.ModTime = info.ModTime()
+
+	r, w := io.Pipe()
+
+	f := &sftpFile{
+		file:   file,
+		reader: r,
+		writer: w,
+		group:  new(errgroup.Group),
+	}
+
+	f.group.Go(f.transfer)
+
+	return f, obj, nil
+}
+
+// sftpFile handles piping the data from a remote sftp file to a consumer. This is mostly here to
+// allow the use of (*sftp.File).WriteTo for reading data, which improves network performance by
+// executing concurrent buffered reads which are fed to the reader while maintaining an accurate
+// streamed representation of the file.
+type sftpFile struct {
+	file   *sftp.File
+	reader *io.PipeReader
+	writer *io.PipeWriter
+	group  *errgroup.Group
+}
+
+func (f *sftpFile) transfer() error {
+	// Pump data to the pipe using WriteTo to utilitize concurrent read workers.
+	_, err := f.file.WriteTo(f.writer)
+
+	// Errors from WriteTo will be propagated to calls to read. CloseWithError always returns `nil`.
+	return f.writer.CloseWithError(err)
+}
+
+func (f *sftpFile) Read(p []byte) (n int, err error) {
+	return f.reader.Read(p)
+}
+
+func (f *sftpFile) Close() error {
+	if err := f.file.Close(); err != nil {
+		return err
+	} else if f.reader.Close(); err != nil {
+		return err
+	}
+
+	return f.group.Wait()
+}
+
+func main() {
+	var src = filesource.Source{
+		NewConfig: func() filesource.Config { return new(config) },
+		Connect: func(ctx context.Context, cfg filesource.Config) (filesource.Store, error) {
+			return newSftpSource(ctx, cfg.(*config))
+		},
+		ConfigSchema:     configSchema,
+		DocumentationURL: "https://go.estuary.dev/source-sftp",
+		// Set the delta to 30 seconds in the past, to guard against new files appearing with a
+		// timestamp that's equal to the `MinBound` in the state.
+		TimeHorizonDelta: time.Second * -30,
+	}
+
+	src.Main()
+}
+
+func configSchema(parserSchema json.RawMessage) json.RawMessage {
+	schema := schemagen.GenerateSchema("SFTP Source", &config{})
+
+	// TODO(whb): This is a little gross but I don't have a better idea for how to embed the parser
+	// schema in the overall schema, and this works.
+	schema.Properties.Set("parser", parserSchema)
+
+	out, err := schema.MarshalJSON()
+	if err != nil {
+		log.Fatal(fmt.Errorf("generating endpoint schema: %w", err))
```

Comment:
Are you intentionally using the `"log"` package here and `logrus` for other logging, or is that accidental? (In either case, this sort of ambiguity is why I prefer consistently referring to the Logrus packge as `log` in new code).

Reply:
Accidental! I'll update to use `logrus` everywhere, aliased as `log`.